File: src/main/java/nihilite/hooks/HookKeys.java
package nihilite.hooks;

import java.util.LinkedHashMap;
import java.util.Map;

/** Canonical method-key builder for the hook-system P0 surface.
 *  Concatenates a class-internal name, method name, and JLS
 *  field descriptor into a single stable string:
 *
 *    <internal-name> "/" <method-name> "#" <descriptor>
 *
 *  The separator `#` is not legal in any JVM-internal name, method
 *  name, or JLS descriptor, so the concatenation is unambiguous.
 *
 *  Mirrors `nihilite.registry/method-key` byte-for-byte; the
 *  shared test in `nihilite.test.hook-keys` proves agreement.
 *  A bounded {@link LinkedHashMap} cache caps the {@link StringBuilder}
 *  churn on the hot class-load path; the cache is FIFO at 1024
 *  entries so repeated triples resolve in O(1) without leaking.
 */
public final class HookKeys {

    /** Bounded FIFO cache of `(internal|name|descriptor) -> methodKey`. */
    private static final int CACHE_SIZE = 1024;
    private static final Map<String, String> CACHE =
            java.util.Collections.synchronizedMap(
                    new LinkedHashMap<String, String>(CACHE_SIZE, 0.75f, true) {
                        @Override
                        protected boolean removeEldestEntry(
                                Map.Entry<String, String> eldest) {
                            return size() > CACHE_SIZE;
                        }
                    });

    private HookKeys() {}

    /** Build the canonical method-key string. Pure function on the
     *  three inputs; identical inputs always yield identical
     *  outputs and the result is interned via the bounded cache.
     */
    public static String build(String internal, String methodName, String descriptor) {
        String cacheKey = internal + "|" + methodName + "|" + descriptor;
        String cached = CACHE.get(cacheKey);
        if (cached != null) return cached;
        String fresh = internal + "/" + methodName + "#" + descriptor;
        CACHE.put(cacheKey, fresh);
        return fresh;
    }
}

File: src/test/clojure/nihilite/test/hook_keys.clj
(ns nihilite.test.hook-keys
  "Method-key shape contract: the Java-side `HookKeys.build`
   and the Clojure-side `nihilite.registry/method-key` MUST
   produce identical strings for the same input triple
   `(class-internal, method-name, descriptor)`. The shape is
   fixed by the P0 plan: `<internal>` + `/` + `<method-name>`
   + `#` + `<descriptor>`. The separator `#` is not legal in
   any JVM-internal name, method name, or JLS descriptor, so
   the concatenation is unambiguous."
  (:require [clojure.test :refer [deftest is]]
            [nihilite.registry])
  (:import [nihilite.hooks HookKeys]))

(def ^:private internal     "net/minecraft/server/MinecraftServer")
(def ^:private method-name  "sendSystemMessage")
(def ^:private descriptor   "(Lnet/minecraft/network/chat/Component;Z)V")
(def ^:private expected-key
  (str internal "/" method-name "#" descriptor))

(deftest java-build-produces-canonical-shape
  (is (some? HookKeys)
      "HookKeys class is on the runtime classpath")
  (is (= expected-key
         (HookKeys/build internal method-name descriptor))
      "HookKeys/build concatenates with the canonical separator"))

(deftest clojure-method-key-produces-canonical-shape
  (is (= expected-key
         (nihilite.registry/method-key internal method-name descriptor))
      "nihilite.registry/method-key concatenates with the canonical separator"))

(deftest java-and-clojure-agree-on-empty-descriptor
  (let [k1 (HookKeys/build "x/y" "m" "()V")
        k2 (nihilite.registry/method-key "x/y" "m" "()V")]
    (is (= "x/y/m#()V" k1)
        "Java build for trivial triple")
    (is (= k1 k2)
        "Java and Clojure agree byte-for-byte on the trivial triple")))

(deftest java-and-clojure-agree-on-long-descriptor
  (let [d "(Lnet/minecraft/network/chat/Component;Z)V"
        k1 (HookKeys/build "net/minecraft/server/MinecraftServer"
                           "sendSystemMessage" d)
        k2 (nihilite.registry/method-key "net/minecraft/server/MinecraftServer"
                                         "sendSystemMessage" d)]
    (is (= k1 k2)
        "Java and Clojure agree byte-for-byte on the canonical MC triple")))